File: oacis/oacis_watcher.py
import time
import logging
import signal
import sys
import os
from functools import reduce
import fibers

class OacisWatcher():

    def __init__(self, polling = 5, logger = None):
        self._polling = polling
        self._observed_parameter_sets = {}
        self._observed_parameter_sets_all = {}
        self._logger = logger or self._default_logger()
        self._signal_received = False

    def async(self, func):
        f = fibers.Fiber( target=func )
        f.switch()

    def await_ps(self, ps):
        f = fibers.Fiber.current()
        def callback(ps):
            ps.reload()
            f.switch(ps)
        self.watch_ps(ps, callback)
        return f.parent.switch()

    def watch_ps(self, ps, callback):
        psid = ps.id().to_s()
        if psid in self._observed_parameter_sets:
            self._observed_parameter_sets[psid].append( callback )
        else:
            self._observed_parameter_sets[psid] = [ callback ]

    def watch_all_ps(self, ps_array, callback):
        sorted_ps_ids = tuple( sorted( [ ps.id().to_s() for ps in ps_array ] ) )
        if sorted_ps_ids in self._observed_parameter_sets_all:
            self._observed_parameter_sets_all[sorted_ps_ids].append( callback )
        else:
            self._observed_parameter_sets_all[sorted_ps_ids] = [ callback ]

    def loop(self):
        def on_sigint(signalnum, frame):
            print("received SIGNAL %d" % signalnum, file=sys.stderr)
            self._signal_received = True
        org_handler = signal.signal( signal.SIGINT, on_sigint)

        try:
            self._logger.info("start polling")
            while True:
                if self._signal_received:
                    break
                executed = True
                while executed:
                    executed = (self._check_completed_ps() or self._check_completed_ps_all())
                if len(self._observed_parameter_sets) == 0 and len(self._observed_parameter_sets_all) == 0:
                    break
                if self._signal_received:
                    break
                self._logger.info("waiting for %d sec" % self._polling)
                time.sleep(self._polling)
            self._logger.info("stop polling. (interrupted=%s)" % self._signal_received)
        finally:
            signal.signal( signal.SIGINT, org_handler )
            if self._signal_received:
                os.kill( os.getpid(), signal.SIGINT)

    def _default_logger(self):
        logger = logging.getLogger(__name__)
        logger.setLevel( logging.INFO )
        logger.propagate = False
        if not logger.handlers:
            ch = logging.StreamHandler()
            ch.setLevel( logging.INFO )
            formatter = logging.Formatter('%(asctime)s - %(name)s - %(levelname)s - %(message)s')
            ch.setFormatter(formatter)
            logger.addHandler(ch)
        return logger

    def _check_completed_ps(self):
        from . import ParameterSet
        executed = False
        watched_ps_ids = list( self._observed_parameter_sets.keys() )
        psids = self._completed_ps_ids( watched_ps_ids )
        
        for psid in psids:
            if self._signal_received:
                break
            ps = ParameterSet.find(psid)
            if len(ps.runs()) == 0:
                self._logger.info("%s has no run" % ps.id().to_s())
            else:
                self._logger.info("calling callback for %s" % ps.id().to_s())
                executed = True
                queue = self._observed_parameter_sets[psid]
                while len(queue) > 0:
                    callback = queue.pop(0)
                    callback( ps )
                    if self._completed( ps.reload() ) == False:
                        break

        empty_psids = [ psid for psid,callbacks in self._observed_parameter_sets.items() if len(callbacks)==0 ]
        for empty_psid in empty_psids:
            self._observed_parameter_sets.pop(empty_psid)

        return executed

    def _check_completed_ps_all(self):
        from . import ParameterSet
        executed = False
        flattened = reduce(lambda x,y: list(x)+list(y), self._observed_parameter_sets_all.keys(), [])
        watched_ps_ids = list( set(flattened) )
        completed = self._completed_ps_ids( watched_ps_ids )

        for psids in list(self._observed_parameter_sets_all.keys()):
            callbacks = self._observed_parameter_sets_all[psids]
            if self._signal_received:
                break
            if all( (psid in completed) for psid in psids ):
                self._logger.info("calling callback for %s" % repr(psids))
                executed = True
                watched_pss = [ ParameterSet.find(psid) for psid in psids ]
                while len(callbacks) > 0:
                    callback = callbacks.pop(0)
                    callback(watched_pss)
                    if any( self._completed(ps.reload()) for ps in watched_pss ):
                        break

        empty_keys = [ key for key,callbacks in self._observed_parameter_sets_all.items() if len(callbacks)==0 ]
        for key in empty_keys:
            self._observed_parameter_sets_all.pop(key)
        return executed

    def _completed_ps_ids(self, watched_ps_ids ):
        from . import Run
        query = Run.send('in',parameter_set_id = watched_ps_ids).send('in',status=['created','submitted','running']).selector()
        incomplete_ps_ids = [ psid.to_s() for psid in Run.collection().distinct( "parameter_set_id", query ) ]
        completed = list( set(watched_ps_ids) - set(incomplete_ps_ids) )
        return completed

    def _completed(self, ps):
        return ps.runs().send('in', status=['created','submitted','running']).count() == 0

File: lib/oacis_watcher.rb
require 'pp'
require 'logger'
require 'fiber'

class OacisWatcher

  @@current = nil

  def self.current
    @@current
  end

  def self.start( logger: Logger.new($stderr), polling: 5 )
    previous_w = @@current
    w = self.new( logger: logger, polling: polling )
    @@current ||= w
    Fiber.new { yield w }.resume
    w.send(:start_polling)
    @@current = previous_w
  end

  attr_reader :logger

  def initialize( logger: Logger.new($stderr), polling: 5 )
    @observed_parameter_sets = {}
    @observed_parameter_sets_all = {}
    @logger = logger
    @polling = polling
  end

  def self.async(&block)
    @@current.async(&block)
  end

  def self.await_ps(ps)
    @@current.await_ps(ps)
  end

  def self.await_all_ps(ps_list)
    @@current.await_all_ps(ps_list)
  end

  def async(&block)
    Fiber.new(&block).resume
  end

  def await_ps(ps)
    f = Fiber.current
    watch_ps(ps) {
      ps.reload
      f.resume ps
    }
    Fiber.yield
  end

  def await_all_ps(ps_array)
    f = Fiber.current
    watch_all_ps(ps_array) {
      ps_array.each {|ps| ps.reload}
      f.resume ps_array
    }
    Fiber.yield
  end

  def watch_ps(ps, &block)
    psid = ps.id
    if @observed_parameter_sets.has_key?(psid)
      @observed_parameter_sets[ psid ].push( block )
    else
      @observed_parameter_sets[ psid ] = [ block ]
    end
  end

  def watch_all_ps( ps_array, &block )
    sorted_ps_ids = ps_array.map(&:id).sort
    if @observed_parameter_sets_all.has_key?(sorted_ps_ids)
      @observed_parameter_sets_all[ sorted_ps_ids ].push( block )
    else
      @observed_parameter_sets_all[ sorted_ps_ids ] = [ block ]
    end
  end

  private
  def start_polling
    @sigint_received = false
    default_sigaction = Signal.trap("INT") {
      $stderr.puts "received SIGINT"
      @sigint_received = true
    }

    @logger.info "start polling"
    loop do
      break if @sigint_received
      begin
        executed = (check_completed_ps || check_completed_ps_all)
      end while executed
      break if @observed_parameter_sets.empty? && @observed_parameter_sets_all.empty?
      break if @sigint_received
      @logger.debug "waiting for #{@polling} sec"
      sleep @polling
    end
    @logger.info "received SIGINT"  if @sigint_received
    @logger.info "stop polling"
  ensure
    Signal.trap("INT", default_sigaction || "DEFAULT")
    Process.kill("INT", 0) if @sigint_received  # send INT to the current process
  end

  def completed?( ps )
    ps.runs.in( status: [:created, :submitted, :running] ).count == 0
  end

  def completed_ps_ids( watched_ps_ids )
    query = Run.in(parameter_set_id: watched_ps_ids).in(status: [:created,:submitted,:running]).selector
    incomplete_ps_ids = Run.collection.distinct( "parameter_set_id", query )
    watched_ps_ids - incomplete_ps_ids
  end

  # return true, if a callback is executed
  def check_completed_ps
    executed = false
    watched_ps_ids = @observed_parameter_sets.keys
    psids = completed_ps_ids( watched_ps_ids )

    psids.each do |psid|
      break if @sigint_received
      ps = ParameterSet.find(psid)
      if ps.runs.count == 0
        @logger.warn "#{ps} has no run"
      else
        @logger.debug "calling callback for #{psid}"
        executed = true
        while callback = @observed_parameter_sets[psid].shift
          callback.call( ps )
          break unless completed?( ps.reload )
        end
      end
    end

    @observed_parameter_sets.delete_if {|psid, procs| procs.empty? }
    executed
  end

  def check_completed_ps_all
    executed = false
    watched_ps_ids = @observed_parameter_sets_all.keys.flatten
    completed = completed_ps_ids( watched_ps_ids )

    @observed_parameter_sets_all.keys.each do |watched_ps_ids|
      callbacks = @observed_parameter_sets_all[watched_ps_ids]
      if watched_ps_ids.all? {|psid| completed.include?(psid) }
        @logger.debug "calling callback for #{watched_ps_ids}"
        executed = true
        while callback = callbacks.shift
          watched_pss = watched_ps_ids.map {|psid| ParameterSet.find(psid) }
          callback.call( watched_pss )
          break if watched_pss.any? {|ps| ! completed?(ps.reload) }
        end
      end
    end

    @observed_parameter_sets_all.delete_if {|psids, procs| procs.empty? }
    executed
  end
end

